lib/edamam.rb

```diff
@@ -0,0 +1,199 @@
+module Edamam
+  API_HOST = 'api.edamam.com'.freeze
+  APP_ID = ENV.fetch('EDAMAM_APP_ID')
+  APP_KEY = ENV.fetch('EDAMAM_APP_KEY')
+  MEASUREMENT_URI_PREFIX = 'http://www.edamam.com/ontologies/edamam.owl#Measure_'.freeze
+  NUTRITION_DATA_PATH = '/api/food-database/nutrients'.freeze
+  UPC_LOOKUP_PATH = '/api/food-database/parser'.freeze
+
+  class << self
+    def get_grams_per_measurement_unit(food_id:, measurement:)
+      data = send_request(
+        body: {
+          ingredients: [
+            {
+              quantity: 1,
+              measureURI: measurement_uri(measurement),
+              foodId: food_id
+            }
+          ]
+        }.to_json,
+        headers: { 'Content-Type': 'application/json' },
+        host: API_HOST,
+        method: :post,
+        name: 'measurement data',
+        path: NUTRITION_DATA_PATH
+      )
+      get_weight_from_nutrition_data_response(data)&.to_d
+    end
+
+    def get_nutrition_data_from_upc(upc)
+      upc_lookup_data = send_request(
+        host: API_HOST,
+        method: :get,
+        name: 'UPC lookup',
+        path: UPC_LOOKUP_PATH,
+        query: { upc: upc }
+      )&.dig('hints', 0)
+
+      return unless upc_lookup_data
+
+      food_data = upc_lookup_data['food']
+      return unless food_data
+
+      food_id = food_data['foodId']
+      name = food_data['label']
+      return unless food_id && name
+
+      measurement_data = upc_lookup_data['measures']
+      measurement_units =
+        if measurement_data
+          measurement_data.map do |measure|
+            match_data = measure['uri'].match(/^#{MEASUREMENT_URI_PREFIX}([a-z]+)$/)
+            next unless match_data
+
+            [measure['label'], match_data[1]]
+          end.compact.to_h
+        else
+          {}
+        end
+
+      body = {
+        ingredients: [
+          {
+            quantity: 1,
+            measureURI: measurement_uri('tablespoon'),
+            foodId: food_id
+          }
+        ]
+      }
+      request_data = {
+        body: body.to_json,
+        headers: { 'Content-Type': 'application/json' },
+        host: API_HOST,
+        method: :post,
+        name: 'nutrition data',
+        path: NUTRITION_DATA_PATH
+      }
+      response = send_request(request_data)
+
+      grams_per_tablespoon = get_weight_from_nutrition_data_response(response)
```

Comment:
Why are we making this request unconditionally? We already know whether tablespoons are a supported measurement.

Reply:
This is just a method to parse the response, not a request.

Reply:
This is just a method to parse the response, not a request.

Reply:
oh wait I get what you mean